### PR TITLE
Beta TC VAE

### PR DESCRIPTION
**Pull request by jaceybronte**
This PR implements the Beta TC VAE and cleans up the train-VAE folder

### CODE REVIEW

File: 2.train-VAE/betatcvae.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
+# betatcvae.py
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from typing import List
+import pandas as pd 
+import numpy as np
+import pathlib
+import math
```
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py]: Consider making these in alphabetical order using PEP8 import standards. 

Might consider using isort with pre-commit : https://github.com/WayScience/MNIST_TD-VAE/blob/master/.pre-commit-config.yaml

File: 2.train-VAE/betatcvae.py [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+        Decodes the latent variable into the reconstructed input.
+    reparameterize(mu: torch.Tensor, logvar: torch.Tensor) -> torch.Tensor
+        Reparameterization trick to sample from the latent space.
+    forward(x: torch.Tensor)
+        Forward pass through the network.
+    log_density_gaussian(x: torch.Tensor, mu: torch.Tensor, logvar: torch.Tensor)
+        Computes the log density of a Gaussian.
+    loss_function(*args, **kwargs) -> dict
+        Computes the loss for the Beta-TCVAE.
+    """
```
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py:63]: Nice docstring! :D

File: 2.train-VAE/betatcvae.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+        # Initialize class attributes
+        self.latent_dim = latent_dim
+        self.anneal_steps = anneal_steps
+        self.alpha = alpha
+        self.beta = beta
+        self.gamma = gamma
+
+        if hidden_dims is None:
+            hidden_dims = [256, 256, 256]  # Default hidden dimensions if none are provided
```
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py]: Consider making this an optional arg with these dimension as the default. This will cut down on lines of code. I will make the suggestion above

File: 2.train-VAE/betatcvae.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+        Computes the loss for the Beta-TCVAE.
+    """
+    
+    num_iter = 0  # Global static variable to keep track of iterations
+    
+    def __init__(self,
+                 input_dim: int,
+                 latent_dim: int,
+                 beta: float,
+                 hidden_dims: List = None,
```
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py]:
```suggestion
                 hidden_dims: List =  [256, 256, 256],
```

File: 2.train-VAE/betatcvae.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+
+        self.encoder = nn.Sequential(*modules)
+        self.fc1 = nn.Linear(input_dim, 256) 
+        self.fc_mu = nn.Linear(256, latent_dim) 
+        self.fc_logvar = nn.Linear(256, latent_dim)  
+        self.fc2 = nn.Linear(latent_dim, 256)  
+        self.fc3 = nn.Linear(256, input_dim) 
+
+        # Build Decoder
+        self.decoder_input = nn.Linear(latent_dim, 256)
```
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py]: Consider not hardcoding the 256 here and using a variable instead

File: 2.train-VAE/betatcvae.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+                    nn.Linear(hidden_dims[i], hidden_dims[i + 1]),
+                    nn.ReLU())
+            )
+
+        self.decoder = nn.Sequential(*modules)
+        self.final_layer = nn.Linear(hidden_dims[-1], input_dim)  
+
+    def encode(self, input: torch.Tensor) -> List[torch.Tensor]:
+        result = self.encoder(input) 
+        print(f"Encoder output shape: {result.shape}")
```
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py]: Consider logging instead of print statements here

Comment (MikeLippincott) [reply 2.train-VAE/betatcvae.py]: Might be an issue for another PR though?

File: 2.train-VAE/betatcvae.py [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+            modules.append(
+                nn.Sequential(
+                    nn.Linear(hidden_dims[i], hidden_dims[i + 1]),
+                    nn.ReLU())
+            )
+
+        self.decoder = nn.Sequential(*modules)
+        self.final_layer = nn.Linear(hidden_dims[-1], input_dim)  
+
+    def encode(self, input: torch.Tensor) -> List[torch.Tensor]:
```
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py:118]: consider adding docstrings to these methods

File: 2.train-VAE/betatcvae.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+
+    def loss_function(self,
+                      *args,
+                      **kwargs) -> dict:
+        recons = args[0]  
+        input = args[1]  
+        mu = args[2]  
+        log_var = args[3]  
+        z = args[4] 
+        M_N = kwargs.get('M_N', 1.0)  # Minibatch size
```
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py]: generally, all caps variables are global vars. Consider refactoring this one.

File: 2.train-VAE/betatcvae.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+            self.beta * tc_loss +
+            anneal_rate * self.gamma * kld_loss)
+        
+        return loss
+
+
+
+
+    
+def train_model(model, train_loader, optimizer):
```
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py:209]: consider adding doc strings and type hints here

File: 2.train-VAE/betatcvae.py [hunk 10]
```diff
@@ -0,0 +1,300 @@
...
+    latent_space = []
+    with torch.no_grad():
+        for batch in data_loader:
+            data = batch[1]
+            mu, log_var = model.encode(data)  # Extract the mean and log variance
+            latent_space.append(mu.cpu().numpy())  # Append the mean to the latent space
+    latent_space = np.concatenate(latent_space, axis=0)
+    latent_df = pd.DataFrame(latent_space)
+    latent_df.insert(0, 'ModelID', metadata['ModelID'])
+    latent_df_dir = pathlib.Path("./results/latent_df_tc.parquet")
```
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py]: consider making this path an arg for the function

Comment (MikeLippincott) [reply 2.train-VAE/betatcvae.py]: We want these functions to be flexible and reusable across analysis, and repos even

File: 2.train-VAE/betatcvae.py [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+    weight_df_T_df = weight_df.T
+    gene_weight_df = pd.DataFrame(data=weight_df_T_df.values, columns=subset_train_df.columns)
+    gene_weight_T_df = gene_weight_df.T
+    gw_reindex_df = gene_weight_T_df.reset_index()
+    gw_renumber_df = gw_reindex_df.rename(columns={x: y for x, y in zip(gw_reindex_df.columns, range(0, len(gw_reindex_df.columns)))})
+    split_data_df = gw_renumber_df[0].str.split(" ", expand=True)
+    gene_name_df = split_data_df.iloc[:, :1]
+    trimmed_gene_weight_df = gw_renumber_df.iloc[:, 1:]
+    final_gene_weights_df = gene_name_df.join(trimmed_gene_weight_df)
+    gene_weight_dir = pathlib.Path("./results/weight_matrix_gsea_tc.parquet")
```
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py]: same here

File: 2.train-VAE/betatcvae.py (hunk 9 above)
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py:209]: Also consider splitting these up into different `.py` files. I could see a `tcvae_model.py` file for the class and a `tcvae_train_utils.py` file for the training functions.

File: 2.train-VAE/betatcvae.py [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+    latent_space = np.concatenate(latent_space, axis=0)
+    latent_df = pd.DataFrame(latent_space)
+    latent_df.insert(0, 'ModelID', metadata['ModelID'])
+    latent_df_dir = pathlib.Path("./results/latent_df_tc.parquet")
+    latent_df.to_parquet(latent_df_dir, index=False)
+    return latent_df
+
+def weights(model, subset_train_df):
+    # Access the first layer of the encoder within the Sequential container
+    weight_matrix = model.encoder[0][0].weight.detach().cpu().numpy().T  # Extract weights from the first linear layer
```
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py:346]: Is there a way to make this programmatic to extract weights from other layers too?

File: 2.train-VAE/betatcvae.py [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+        recon_x = self.fc3(h2)  # Get reconstruction
+        return recon_x, mu, logvar, z
+
+    def log_density_gaussian(self, x: torch.Tensor, mu: torch.Tensor, logvar: torch.Tensor):
+        normalization = -0.5 * (math.log(2 * math.pi) + logvar) 
+        inv_var = torch.exp(-logvar)  
+        log_density = normalization - 0.5 * ((x - mu)**2 * inv_var)  
+        return log_density
+
+    def loss_function(self,
```
Comment (MikeLippincott) [inline 2.train-VAE/betatcvae.py:151]: This is a complex loss function. I think some inline documentation might be useful for others to understand better. Consider adding comments explaining what each term is and what it is doing.

File: 2.train-VAE/optimize_utils_tcvae.py [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+        "learning_rate", args.min_lr, args.max_lr
+    )
+    batch_size = trial.suggest_int(
+        "batch_size", args.min_batch_size, args.max_batch_size
+    )
+    epochs = trial.suggest_int(
+        "epochs", args.min_epochs, args.max_epochs
+    )
+    optimizer_type = trial.suggest_categorical(
+        "optimizer_type", ["adam","rmsprop"]
```
Comment (MikeLippincott) [inline 2.train-VAE/optimize_utils_tcvae.py:180]: No SGD?

File: 2.train-VAE/scripts/3.beta_tc_vae_training_extract_model_components .py [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+import optuna
+import matplotlib.pyplot as plt
+from torch.utils.data import DataLoader, TensorDataset
+from pathlib import Path
+from betatcvae import BetaTCVAE, compile_vae, extract_latent_dimensions, weights
+from optimize_utils_tcvae import get_optimizer
+
+script_directory = pathlib.Path("../utils/").resolve()
+sys.path.insert(0, str(script_directory))
+from data_loader import load_train_test_data
```
Comment (MikeLippincott) [inline 2.train-VAE/scripts/3.beta_tc_vae_training_extract_model_components .py:23]: same here, consider using isort

File: 2.train-VAE/scripts/3.beta_tc_vae_training_extract_model_components .py [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+learning_rate = best_trial.params['learning_rate']
+batch_size = best_trial.params['batch_size']
+epochs = best_trial.params['epochs']
+optimizer = best_trial.params['optimizer_type']
+
+# Create DataLoader
+train_loader = DataLoader(TensorDataset(train_tensor), batch_size=batch_size, shuffle=True)
+val_loader = DataLoader(TensorDataset(val_tensor), batch_size=batch_size, shuffle=False)
+test_loader = DataLoader(TensorDataset(test_tensor), batch_size=batch_size, shuffle=False)
+
```
Comment (MikeLippincott) [inline 2.train-VAE/scripts/3.beta_tc_vae_training_extract_model_components .py:60]: nice!

**Review by MikeLippincott — APPROVED** [review]
Good job, ready to merge after addressing comments. Let me know if you want help setting up pre-commit! LGTM.

File: 2.train-VAE/betatcvae.py (hunk 1 above)
Comment (jaceybronte) [reply 2.train-VAE/betatcvae.py]: Fixed

File: 2.train-VAE/betatcvae.py (hunk 2 above)
Comment (jaceybronte) [reply 2.train-VAE/betatcvae.py:63]: Thanks!

File: 2.train-VAE/betatcvae.py (hunk 4 above)
Comment (jaceybronte) [reply 2.train-VAE/betatcvae.py]: fixed

File: 2.train-VAE/betatcvae.py (hunk 3 above)
Comment (jaceybronte) [reply 2.train-VAE/betatcvae.py]: fixed

File: 2.train-VAE/betatcvae.py (hunk 5 above)
Comment (jaceybronte) [reply 2.train-VAE/betatcvae.py]: fixed

File: 2.train-VAE/betatcvae.py (hunk 6 above)
Comment (jaceybronte) [reply 2.train-VAE/betatcvae.py]: this was just to find a matrix multiplication error, removed!

File: 2.train-VAE/betatcvae.py (hunk 7 above)
Comment (jaceybronte) [reply 2.train-VAE/betatcvae.py:118]: done

File: 2.train-VAE/betatcvae.py (hunk 8 above)
Comment (jaceybronte) [reply 2.train-VAE/betatcvae.py]: fixed

File: 2.train-VAE/betatcvae.py (hunk 9 above)
Comment (jaceybronte) [reply 2.train-VAE/betatcvae.py:209]: fixed

File: 2.train-VAE/betatcvae.py (hunk 10 above)
Comment (jaceybronte) [reply 2.train-VAE/betatcvae.py]: done

File: 2.train-VAE/betatcvae.py (hunk 11 above)
Comment (jaceybronte) [reply 2.train-VAE/betatcvae.py]: done

File: 2.train-VAE/scripts/3.beta_tc_vae_training_extract_model_components .py (hunk 15 above)
Comment (jaceybronte) [reply 2.train-VAE/scripts/3.beta_tc_vae_training_extract_model_components .py:23]: fixed

File: 2.train-VAE/optimize_utils_tcvae.py (hunk 14 above)
Comment (jaceybronte) [reply 2.train-VAE/optimize_utils_tcvae.py:180]: just not giving any sort of good results and adjusting other hyperparams to fix has negative effects on adam and rmsprop

File: 2.train-VAE/betatcvae.py (hunk 13 above)
Comment (jaceybronte) [reply 2.train-VAE/betatcvae.py:151]: yep
